Issue 223/delete client modal scott

xscottxbrownx opened the pull request:
# This PR closes #223 

This PR:
1. Creates `DeleteClientModal` component.
2. Renders this component onClick of trash icon in `ClientListTableRow`.
3. Renames states of `AddClientModal` component to achieve more clarity.
4. Automatically closes the `AddClientModal` component once client has been added to the list.

---


https://github.com/codeforpdx/PASS/assets/71395931/071d91d4-f4a9-4745-82b2-294d106289d2

xscottxbrownx committed:
renaming states (due to another modal being added), and setting modal to auto-close

xscottxbrownx committed:
create DeleteClientModal.jsx to replace alert pop-up

andycwilliams commented on src/components/Clients/ClientListTableRow.jsx:
```diff
@@ -65,45 +62,55 @@ const ClientListTableRow = ({ labelId, client, state, dispatch }) => {
-        </IconButton>
-      </StyledTableCell>
-    </StyledTableRow>
+    <>
```
Not critical at all, but empty fragments strike me as unusual. However, the component needs restyling anyway so this wouldn't even need to be addressed now.

xscottxbrownx replied:
React components can only return one thing, so it was necessary because I was returning the `<StyledTableRow>` and the `<DeleteClientModal>`.  The fragments don't produce an actual dom element so it's not in the user facing code, just here in the dev side of things - which I feel is better than using a `<div>` to accomplish which would be in the DOM at the end.

But with all that said, could put the modal inside the table row so no need for react fragments, but even though it would work fine I felt like this is a separate component from the table row so I separated it. 

I'm up for switching it, if you all feel like it should be switched no problem.

leekahung replied:
We do have two separate components in here after all (one for the row, one for the modal). I think wrapping this in a fragment is appropriate.

andycwilliams commented on src/components/Clients/DeleteClientModal.jsx:
```diff
@@ -0,0 +1,64 @@
+
+    <DialogContent>
+      <DialogContentText id="dialog-description">
+        {`You're about to delete ${client.person} from your client list, do you wish to continue?`}
```
Minor point, but I notice in the demonstration video that there's a bit of extra spacing on the right side of the modal. That is, "to continue" is wrapped around but the "DELETE CLIENT" button stretches the modal out further to the right. It might be better if the edge of the buttons is made to align with the edge of the rightmost text, wherever that is.

xscottxbrownx replied:
Interesting, I didn't notice that previously...

Looks like this modal, the Logout Modal, and the Inactivity Message are all this way. Like the dialog buttons are always over on the right, even if the sx styling is removed.  Have to look at theme or how to override for this.

andycwilliams reviewed (approved):
Approving these changes since my comments are not essential. The code is clean and addresses the related issue.

leekahung commented on src/components/Clients/ClientListTableRow.jsx:
```diff
@@ -46,13 +49,7 @@ const ClientListTableRow = ({ labelId, client, state, dispatch }) => {
-    ) {
-      return;
-    }
+    setShowClientDeleteModal(false);
```
Not really an issue per-se since I believe this is appropriate too. But the modal seems to close even without this line when I click on "Delete Client".

https://github.com/codeforpdx/PASS/assets/14917816/f6839acb-401f-4fc0-8526-9d2e0f84ab10

I wonder if it has something to do with the client being removed from the list so the modal disappears on its own. If it is because of that reason, I think keeping this line should be fine since an error for whatever reason would cause the modal stuck being opened.

xscottxbrownx replied:
It closes when the action is complete.  I was having it close early, so user could see the status notifications that are displayed on the table.

Could move the notifications to the modal, as Tim said below.  

Or have notifications some other way:
 - Just a spinner while it's running (https://mui.com/material-ui/react-progress/#circular) and then a snackbar for error or success (https://mui.com/material-ui/react-snackbar/#customization)
 - Interactive integration (https://mui.com/material-ui/react-progress/#interactive-integration)

xscottxbrownx replied:
@timbot1789 since you mentioned it in another comment, take a look at these options as well and see what you think.

timbot1789 replied:
I like the snackbar because we'll probably want to use something similar for other messages throughout the system.

andycwilliams replied:
This is related to issue #227 and it looks like Russell just assigned himself to it

leekahung commented on src/components/Clients/ClientListTableRow.jsx:
```diff
@@ -65,45 +62,55 @@ const ClientListTableRow = ({ labelId, client, state, dispatch }) => {
-      </StyledTableCell>
-    </StyledTableRow>
+    <>
+      <StyledTableRow>
```
Oh, one thing I want to touch, do we want to start moving these styled-components over to Material UI now, or have it in a separate PR?

andycwilliams replied:
My vote would be to do it now, but am also okay with leaving for another PR.

xscottxbrownx replied:
That's Issue #219 assigned to Masen. Pinged them in our Discord DMs.

leekahung commented:
My bad, I messed up updating the wrong branch. I've reverted the changes back to your latest commit

timbot1789 commented on src/components/Clients/ClientListTableRow.jsx:
```diff
@@ -65,45 +62,55 @@ const ClientListTableRow = ({ labelId, client, state, dispatch }) => {
+      </StyledTableRow>
+
+      {/* modal/popup renders when showDeleteClientModal state is true */}
+      <DeleteClientModal
```
It makes more sense to me to put this on the Table or Page level. That way we won't be rendering 100 different modals for 100 different rows. It would also resolve the `React.Fragment` issue that Andy mentioned higher up.

You can then have the `handleDeleteClient` function call some passed-in messaging function that triggers the modal, like 

`const handleDeleteClient = () => props.callDelete(client)`

If you don't want to use props, you could set up a reducer and have the delete fire off an action, but that will be hard to integrate with the context. I haven't really thought about how to integrate contexts with reducers, so I don't have much advice there.

timbot1789 reviewed (commented):
Could you add a snapshot test for the modal? There are some example snapshots in the `tests` folder, and I can go over writing them with you if you like.

timbot1789 commented on src/components/Clients/ClientListTableRow.jsx:
```diff
@@ -46,13 +49,7 @@ const ClientListTableRow = ({ labelId, client, state, dispatch }) => {
-      return;
-    }
+    setShowClientDeleteModal(false);
     runNotification(`Deleting "${client.person}" from client list...`, 3, state, dispatch);
```
Non-Issue: Any thoughts on moving these notifications into the modal? Seems like if we're going to be displaying messages about deletion, we can display them all in one place. Doesn't have to be in this PR if we want to change it.

xscottxbrownx replied:
See comment above from Ka Hung and my quickly researched response.

xscottxbrownx committed:
Adding DeleteClientModal to /Clients/index.js

xscottxbrownx committed:
Adding all client components to its index.js

xscottxbrownx committed:
moving status notifications to be inside the DeleteClientModal.jsx component

xscottxbrownx committed:
moving DeleteClientModal from table row to just the table

xscottxbrownx committed:
Lifting DeleteClientModal and state up to parent component

xscottxbrownx committed:
edit DeleteClientModal text

xscottxbrownx committed:
Merge branch 'Development' into issue-223/delete-client-modal-scott

xscottxbrownx commented:
OK,

summary:

1. Renamed states for the `AddClientModal` to that instead of `showModal` - due to having another modal in `Clients` now (`DeleteClientModal`) just to keep code more clear.
2. Lifted the `DeleteClientModal` component up to `ClientList` (now contains the `ClientListTable` and this `DeleteClientModal`.)
3. Moved the status notifications to the bottom of the `DeleteClientModal` instead of the `ClientListTable`. (This is still planned to be updated to snackbars from MUI in a separate PR/Issue.)
4. Added all the `Clients` components to its `index` file.
5. Added test file for this component.

timbot1789 committed:
Add basic test cases for Scott to finish

xscottxbrownx commented:
Thanks to @timbot1789 for giving a crash course in testing!

### I'm still having issues trying to pass my last test that I'd like to implement for this PR. So, I will describe what it is and the approaches I have tried so far - maybe somebody can see the issue, or think of another approach to try.

---

### I want to test that the modal does close without taking any action, when the `Cancel` button is pressed.

---

#### 1) Thought of checking to see if specific elements (that are only in the modal) exist after button click (this one was specifically the delete button.  I tried it two different ways:
<img width="701" alt="Screenshot 2023-07-10 at 9 11 59 AM" src="https://github.com/codeforpdx/PASS/assets/71395931/9843f105-9d44-4254-b7f2-4f16ab4cd05c">
<img width="717" alt="Screenshot 2023-07-10 at 9 10 51 AM" src="https://github.com/codeforpdx/PASS/assets/71395931/e53a193a-6fc2-438b-a2ce-23fe4a4951da">

--- 

#### 2) Thought of checking to see if the setter function for the state (that controls whether the modal is displayed or not) was fired on button click:
<img width="709" alt="Screenshot 2023-07-10 at 9 16 04 AM" src="https://github.com/codeforpdx/PASS/assets/71395931/77909adb-685c-42f8-b553-1c9f636a76c7">

I believe approach 2 to be decent, but it is most likely not written correctly. I don't know how it can correlate my made up `closeModal` function to the modal actually calling `setShowDeleteClientModal` state setter function.

---

If anybody sees any issues with the code, has thoughts on a different approach, etc - please ping me with response.

xscottxbrownx committed:
adding aria label to buttons

timbot1789 commented:
@xscottxbrownx Another thing you could do is write a simple wrapper component to launch the modal in, and test that:

```
import { render, cleanup, waitForElementToBeRemoved } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React, { useState } from 'react';
import { expect, it, vi, afterEach } from 'vitest';
import { DeleteClientModal } from '../../../src/components/Clients';
import { UserListContext } from '../../../src/contexts';

...


it('closes without an action when cancel button in clicked', async () => {
  const user = userEvent.setup();
  const client = { person: 'tim' }
  const removeUser = vi.fn();
  const ModalWrapper = () => {
    const [showModal, setShowModal] = useState(true)

    return <UserListContext.Provider value={{ removeUser }}>
      <DeleteClientModal
        showDeleteClientModal={showModal}
        setShowDeleteClientModal={setShowModal}
        selectedClientToDelete={client}
      />
    </UserListContext.Provider>
  }

  const { queryByLabelText } = render(<ModalWrapper/>);

  const cancelButton = queryByLabelText("Cancel Delete Client Modal");
  await user.click(cancelButton);
  // wait for the dialog to drop
  await waitForElementToBeRemoved(() => queryByLabelText("Cancel Delete Client Modal"))
  const hiddenCancelButton = queryByLabelText("Cancel Delete Client Modal");
  expect(hiddenCancelButton).toBeNull();
  expect(removeUser).not.toBeCalled();
});
```
This test seems to work without issue. One thing I had to look up was the `waitForElementToBeRemoved` function, which does about what it says on the box.

---
However, the React Testing Library docs do include a set of examples, one of which is testing modals:
https://testing-library.com/docs/example-react-modal/

You can see that they do the same test as you wrote. So I'm fine with either approach, but you're right, the test would be more robust if it checks that the modal is actually dropped. Not that it just calls the `drop` function.

xscottxbrownx committed:
passing unmount test

xscottxbrownx committed:
having eslint ignore issue in test file

xscottxbrownx committed:
Merge branch 'Development' into issue-223/delete-client-modal-scott

xscottxbrownx committed:
Moving DeleteClientModal so it stays rendered even when ClientList is empty.

xscottxbrownx committed:
making consistent status notification timing on both client modals

leekahung reviewed (approved):
The changes seems to work fine and the code is running well. I'll approve this.

timbot1789 commented on src/components/Clients/DeleteClientModal.jsx:
```diff
@@ -0,0 +1,109 @@
+      state,
+      dispatch
+    );
+    try {
```
This block seems odd to me. It says "attempt to delete the client, then, regardless of if the attempt succeeds or fails, tell the client that it succeeded." I think what you mean is something more like this:

```
try {
  await removeUser();
  runNotification("User deleted!", 5, state, dispatch)
} catch (e) {
  runNotification(`User Deletion failed. Reason: ${e.message}`)
} finally {
  setTimeout(() => {setShowDeleteClientModal(false)}, 2000)
}
```

That says "try to delete the user. If you succeed, display a success message. If you fail, display an error message. No matter what, drop the modal after 2 seconds"

xscottxbrownx replied:
oh yes, agree it needs a `catch` block.

But leaving this before the `try` block:
`runNotification(
      'Deleting "${selectedClientToDelete?.person}" from client list...',
      5,
      state,
      dispatch
    );`
    so the user has some sort of indication that something is going on.

timbot1789 replied:
Works for me!

timbot1789 commented on src/components/Clients/DeleteClientModal.jsx:
```diff
@@ -0,0 +1,109 @@
+      aria-describedby="dialog-description"
+      onClose={() => setShowDeleteClientModal(false)}
+    >
+      <FormSection
```
Followup: This shouldn't really be a form. I get it's because you want to use the notifications system, but that seems to me like a sign that we need a better notifications system. We should be able to send notifications from any kind of component, not just forms. After we rework notifications to be robust, we should consider reworking this form too. That's for the future though.

xscottxbrownx replied:
Fully agreed. I thought we already had an issue for that, but I guess not. I will create one.

Revamping the UI/UX of `<StatusNotification>` is definitely out of the scope of this PR.  I just made the code here the same as it was when the notifications were showing in the client table (and it's how they are coded in the `AddClientModal`, and all the other forms we still have - so it's easier for whichever dev wants to tackle that issue.

timbot1789 replied:
Yup I totally agree that what you did here was the right thing given the current state of the code base. Mostly just wanted to call it out, since we don't want this design pattern to proliferate too much. Nice work either way

xscottxbrownx committed:
adding a catch block for possible errors on deletion of client

andycwilliams commented:
I know this just closed, but I'm running from the latest Dev branch and am seeing this:

![2023-07-11 (2)](https://github.com/codeforpdx/PASS/assets/83156697/5f101adb-ddd8-4e7d-8ccf-c5b5a551d97b)

I haven't had time to check what's causing this yet.

xscottxbrownx commented:
> I know this just closed, but I'm running from the latest Dev branch and am seeing this:
> 
> ![2023-07-11 (2)](https://user-images.githubusercontent.com/83156697/252813895-5f101adb-ddd8-4e7d-8ccf-c5b5a551d97b.png)
> 
> I haven't had time to check what's causing this yet.

Sorry, causing what?  I'm not seeing what you are pointing out.

andycwilliams commented:
I meant the uncentered buttons. Actually I just realized we have a couple of different philosophies with button placement. On some modals they're centered and on others they're not. Not something we need to address now, if at all.

timbot1789 commented:
> I meant the uncentered buttons. Actually I just realized we have a couple of different philosophies with button placement. On some modals they're centered and on others they're not. Not something we need to address now, if at all.

Sounds like a great first issue for someone new to the project.

xscottxbrownx commented:
Yep, see our convo above... it's basically the modals that are just using `<Dialog>` and `<DialogActions>` compared to those that have a `<form>` or `<FormSection>` inside of them.
